### PR TITLE
[PERF] free all indexes on a thread

### PR DESCRIPTION
**Pull request by ashtul**
Improve flow when all indexes are freed using `Indexes_Free`.

`IndexSpec_FreeInternals` is split into two parts - `FreeGlobals` and `FreeInternals` which are used together to Drop a single index. 
When all indexes are free, to avoid a long hang, indexes are first unlinked from all global lists, then a list of all the specs is passed to a thread that frees their internals.

Flow:
1. Unlink the indexes by removing them from all global lists including:
* specs
* spec schema prefixes
* spec aliases
* cursors
* GC

2. On a thread, free resource for all indexes with FreeInternals including:
* schema rule
* indexer
* text terms trie
* trie for field types TAG, NUMERIC and GEO
* doc-table
* sortables
* stopwords
* synonym
* field specs cache
* formated field strings,
* field specs
* scanner

### CODE REVIEW

**Review by MeirShpilraien — COMMENTED** [review]
Please give a better commit message that explains the change:
1. what the change does?
2. what is comes to solve
3. is there any new configuration that was introduce with this change? Should we consider such configuration?
4. Anything else that the reviewer should focus on

**Comment by ashtul** [comment]
> Please give a better commit message that explains the change:
> 
> 1. what the change does?
> 2. what is comes to solve
> 3. is there any new configuration that was introduce with this change? Should we consider such configuration?
> 4. Anything else that the reviewer should focus on

@MeirShpilraien 
1. The change is for the `Indexes_Free` function. When the function is called, it collects all index specs pointers to an array, frees all global structures, and passes the specs array to a thread to free off the main thread.
2. The change improves performance and solves a case of timeout a client experienced.
3. No.
4. `Indexes_Free` function.

File: src/alias.c [hunk 1]
```diff
@@ -24,6 +24,13 @@ void IndexAlias_DestroyGlobal(AliasTable **t) {
   *t = NULL;
 }
 
+void IndexAlias_Empty(AliasTable *t) {
+  if (!t) {
+    return;
```
Comment (MeirShpilraien) [inline src/alias.c:29]: Is this code reachable?

File: src/rdb.c [hunk 2]
```diff
@@ -3,41 +3,44 @@
 #include "rdb.h"
 
 dict *specDict_g_bkup;
-TrieMap *ScemaPrefixes_g_bkup;
+TrieMap *SchemaPrefixes_g_bkup;
```
Comment (MeirShpilraien) [inline src/rdb.c:6]: Is this renaming is part of this PR changes or is it unrelated? If its unrelated please avoid it next time as it make it much harder to concentrate on the relevant parts while review.

File: src/spec.c [hunk 3]
```diff
@@ -1148,20 +1163,35 @@ void IndexSpec_FreeSync(IndexSpec *spec) {
 
 //---------------------------------------------------------------------------------------------
 
-void Indexes_Free(dict *d) {
-  arrayof(IndexSpec *) specs = array_new(IndexSpec *, 10);
+void Indexes_Free(dict *d, TrieMap *schemaPrefixes, void *aliases, void *cursors) {
+  if (!cleanPool) {
+    cleanPool = thpool_init(1);
```
Comment (MeirShpilraien) [inline src/spec.c:1168]: Why not initialise it on module load?

File: src/spec.c [hunk 4]
```diff
@@ -1148,20 +1163,35 @@ void IndexSpec_FreeSync(IndexSpec *spec) {
 
 //---------------------------------------------------------------------------------------------
 
-void Indexes_Free(dict *d) {
-  arrayof(IndexSpec *) specs = array_new(IndexSpec *, 10);
+void Indexes_Free(dict *d, TrieMap *schemaPrefixes, void *aliases, void *cursors) {
```
Comment (MeirShpilraien) [inline src/spec.c:1166]: Why void*?

File: src/spec.c [hunk 5]
```diff
@@ -1100,6 +1099,22 @@ static void IndexSpec_FreeTask(IndexSpec *spec) {
   RedisModule_FreeThreadSafeContext(threadCtx);
 }
 
+static void IndexSpec_FreeAllTask(IndexSpec **specs) {
+  RedisModuleCtx *threadCtx = RedisModule_GetThreadSafeContext(NULL);
```
Comment (MeirShpilraien) [inline src/spec.c:1103]: Don't we have a static ctx?

File: src/spec.c [hunk 6]
```diff
@@ -1100,6 +1099,22 @@ static void IndexSpec_FreeTask(IndexSpec *spec) {
   RedisModule_FreeThreadSafeContext(threadCtx);
 }
 
+static void IndexSpec_FreeAllTask(IndexSpec **specs) {
```
Comment (MeirShpilraien) [inline src/spec.c:1102]: This is an array right? Mark it as one.

File: src/spec.c [hunk 7]
```diff
@@ -992,7 +992,7 @@ char *IndexSpec_GetRandomTerm(IndexSpec *sp, size_t sampleSize) {
 
 ///////////////////////////////////////////////////////////////////////////////////////////////
 
-void IndexSpec_FreeInternals(IndexSpec *spec) {
+void IndexSpec_FreeGlobals(IndexSpec *spec) {
```
Comment (MeirShpilraien) [inline src/spec.c:995]: Can you explain this function and how is it different from `IndexSpec_FreeInternals`. When each of them should be used?

**Review by MeirShpilraien — COMMENTED** [review]
@ashtul please take a look at the comments, my main concern is that you actually take the GIL on the background task so not sure we improve anything. Also I am afraid about concurrency issues, for example you remove the indexes from the global holders but not from GC? What will happened if the GC will run in between.

I also would like to see the much more details top comments with much detailed explanations:
1. List things that is freed on async task.
2. List things that is freed on spot.
3. List of places that you call those free functions (is it always go to the background? do we want it to always go to the background?)

File: src/spec.c (hunk 3 above)
Comment (ashtul) [reply src/spec.c:1168]: We use the same pool with IndexSpec_Free for temporary indexes. Should we change on both?

File: src/spec.c [hunk 8]
```diff
@@ -1672,6 +1702,7 @@ void ReindexPool_ThreadPoolDestroy() {
   if (reindexPool != NULL) {
     RedisModule_ThreadSafeContextUnlock(RSDummyContext);
     thpool_destroy(reindexPool);
+    thpool_destroy(cleanPool);      // allow cleaning thread to finish
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/spec.c:1705]: Yes. `thpool_destroy` protects against it internally.

File: src/spec.c (hunk 4 above)
Comment (ashtul) [reply src/spec.c:1166]: Due to circular include of header files.

File: src/spec.c [hunk 9]
```diff
@@ -1100,6 +1099,22 @@ static void IndexSpec_FreeTask(IndexSpec *spec) {
   RedisModule_FreeThreadSafeContext(threadCtx);
 }
 
+static void IndexSpec_FreeAllTask(IndexSpec **specs) {
+  RedisModuleCtx *threadCtx = RedisModule_GetThreadSafeContext(NULL);
+
+  for (size_t i = 0; i < array_len(specs); ++i) {
+    RedisModule_ThreadSafeContextLock(threadCtx);
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/spec.c:1106]: You are so right!!!

File: src/spec.c (hunk 6 above)
Comment (ashtul) [reply src/spec.c:1102]: How?

File: src/spec.c [hunk 10]
```diff
@@ -1100,6 +1099,22 @@ static void IndexSpec_FreeTask(IndexSpec *spec) {
...
+    // Only free internals
+    IndexSpec_FreeInternals(specs[i]);
+    RedisModule_ThreadSafeContextUnlock(threadCtx);
+    sched_yield();
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/spec.c:1110]: removed

File: src/spec.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (ashtul) [reply src/spec.c:1106]: Tested now without and we are failing. The main idea was not necessarily to not take the lock but not to lock it for an extended period of time and get a timeout on the watchdog.

File: src/spec.c (hunk 6 above)
Comment (MeirShpilraien) [reply src/spec.c:1107]: arrayof

File: src/spec.c [hunk 11]
```diff
@@ -1089,15 +1095,21 @@ static void IndexSpec_FreeTask(IndexSpec *spec) {
   RedisModule_Log(NULL, "notice", "Freeing index %s in background", spec->name);
 #endif
 
-  RedisModuleCtx *threadCtx = RedisModule_GetThreadSafeContext(NULL);
-  RedisModule_AutoMemory(threadCtx);
-  RedisModule_ThreadSafeContextLock(threadCtx);
+  RedisModule_ThreadSafeContextLock(RSDummyContext);
+  RedisModule_AutoMemory(RSDummyContext);
```
Comment (MeirShpilraien) [inline src/spec.c:1099]: Again I guess its not related to this PR, lets avoid it in the future and make it in its own PR.

File: src/spec.c [hunk 12]
```diff
@@ -1148,20 +1160,41 @@ void IndexSpec_FreeSync(IndexSpec *spec) {
 
 //---------------------------------------------------------------------------------------------
 
-void Indexes_Free(dict *d) {
-  arrayof(IndexSpec *) specs = array_new(IndexSpec *, 10);
-  dictIterator *iter = dictGetIterator(d);
+void Indexes_Free(dict *specsDict, TrieMap *schemaPrefixes, void *aliases, void *cursors) {
```
Comment (MeirShpilraien) [inline src/spec.c:1163]: Still void*?

File: src/spec.c [hunk 13]
```diff
@@ -1124,7 +1136,7 @@ void IndexSpec_Free(IndexSpec *spec) {
     thpool_add_work(cleanPool, (thpool_proc)IndexSpec_FreeTask, spec);
     return;
   }
-
+  IndexSpec_FreeGlobals(spec);
   IndexSpec_FreeInternals(spec);
```
Comment (MeirShpilraien) [inline src/spec.c:1140]: So a single Index release is not going to the background?

File: src/spec.c [hunk 14]
```diff
@@ -992,7 +992,10 @@ char *IndexSpec_GetRandomTerm(IndexSpec *sp, size_t sampleSize) {
 
 ///////////////////////////////////////////////////////////////////////////////////////////////
 
-void IndexSpec_FreeInternals(IndexSpec *spec) {
+
+// unlink spec from global structures:
+// specs, schema prefixes, aliases, cursors, and GC 
+void IndexSpec_FreeGlobals(IndexSpec *spec) {
   if (spec->name && dictFetchValue(specDict_g, spec->name) == spec) {
```
Comment (MeirShpilraien) [inline src/spec.c:999]: So this is still looking at `specDict_g` which means that we did not fix the swapdb bug?

File: src/spec.c [hunk 15]
```diff
@@ -1148,20 +1160,41 @@ void IndexSpec_FreeSync(IndexSpec *spec) {
...
+      RedisModule_StopTimer(RSDummyContext, sp->timerId, NULL);
+      sp->isTimerSet = false;
+    }
+    if (cursors && sp->uniqueId) {
+      Cursors_PurgeWithName(cursors, sp->name);
+      CursorList_RemoveSpec(cursors, sp->name);
+    }
+    if (sp->gc) {
+      GCContext_Stop(sp->gc);
+    }
```
Comment (MeirShpilraien) [inline src/spec.c:1188]: I understand we are not calling IndexSpec_FreeGlobals here on purpose (its easier to clean the global structures), let document the function and explain it.

**Review by MeirShpilraien — COMMENTED** [review]
@ashtul added some more comments, so if I understand correctly we only go to the background when flushing all the indexes, is that correct? And if so than why not going to the background even on a single index free? Also I believe that maybe we can have some heuristics of when it is worth going to the background depends on the index size?

In addition, I still see that on  `IndexSpec_FreeGlobals` we still refer to the globals, so we did not fix the bug on swapdb?

File: src/spec.c (hunk 14 above)
Comment (ashtul) [reply src/spec.c:999]: The swapdb uses the `Indexes_Free` function which does not call `IndexSpec_FreeGlobals`

**Comment by ashtul** [comment]
replaced by #2560

**Pull request closed without merging**